### PR TITLE
[Enhancement] iceberg partition pruning support transformed partitions

### PR DESCRIPTION
**Pull request by stephen-shelby**
Fixes #issue

## What type of PR is this:

- [ ] BugFix
- [ ] Feature
- [x] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

Does this PR entail a change in behavior?

- [ ] Yes, this PR will result in a change in behavior.
- [x] No, this PR will not result in a change in behavior.

If yes, please specify the type of change:

- [ ] Interface/UI changes: syntax, type conversion, expression evaluation, display information
- [ ] Parameter changes: default values, similar parameters but with different default values
- [ ] Policy changes: use new policy to replace old one, functionality automatically enabled
- [ ] Feature removed
- [ ] Miscellaneous: upgrade & downgrade compatibility, etc.

## Checklist:

- [ ] I have added test cases for my bug fix or my new feature
- [ ] This pr needs user documentation (for new or modified features or behaviors)
  - [ ] I have added documentation for my new feature or new function

## Bugfix cherry-pick branch check:

- [x] I have checked the version labels which the pr will be auto-backported to the target branch
  - [ ] 3.2
  - [ ] 3.1
  - [ ] 3.0
  - [ ] 2.5

### CODE REVIEW

**Comment by wanpengfei-git** [comment]
#### [BE Incremental Coverage Report]


:heart_eyes: **pass** : 0 / 0 (0%)

File: fe/fe-core/src/main/java/com/starrocks/connector/iceberg/IcebergMetadata.java [hunk 1]
```diff
@@ -309,9 +310,14 @@ public List<PartitionKey> getPrunedPartitions(Table table, ScalarOperator predic
 
             StructLike partitionData = fileScanTask.file().partition();
             List<String> values = PartitionUtil.getIcebergPartitionValues(spec, partitionData);
+            if (values.size() != partitionColumns.size()) {
+                // ban partition evolution and non-identify column.
+                continue;
+            }
             try {
-                partitionKeys.add(createPartitionKey(values, icebergTable.getPartitionColumns(), table.getType()));
+                partitionKeys.add(createPartitionKey(values, partitionColumns, table.getType()));
```
Comment (zombee0) [inline fe/fe-core/src/main/java/com/starrocks/connector/iceberg/IcebergMetadata.java:318]: for transformed partition column, partitionColumn get the base column, and value get data from partition data(transformed), is that ok?

**Comment by wanpengfei-git** [comment]
#### [BE Incremental Coverage Report]


:heart_eyes: **pass** : 0 / 0 (0%)

File: fe/fe-core/src/main/java/com/starrocks/connector/iceberg/IcebergMetadata.java (hunk 1 above)
Comment (stephen-shelby) [reply fe/fe-core/src/main/java/com/starrocks/connector/iceberg/IcebergMetadata.java:324]: At present, it seems that this is equivalent. maybe when we really use partitionkey, we need to make adjustments here.

**Review by gengjun-git — APPROVED** [review]

**Comment by wanpengfei-git** [comment]
#### [FE Incremental Coverage Report]

:heart_eyes: **pass** : 17 / 21 (80.95%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/connector/iceberg/IcebergMetadata.java|3|5|60.00%|[321, 326]|
|:large_blue_circle:|com/starrocks/connector/PartitionUtil.java|4|5|80.00%|[632]|
|:large_blue_circle:|com/starrocks/catalog/IcebergTable.java|10|11|90.91%|[155]|

**Comment by wanpengfei-git** [comment]
#### [BE Incremental Coverage Report]


:heart_eyes: **pass** : 0 / 0 (0%)

**Comment by wanpengfei-git** [comment]
#### [FE Incremental Coverage Report]

:heart_eyes: **pass** : 17 / 21 (80.95%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/connector/iceberg/IcebergMetadata.java|3|5|60.00%|[321, 326]|
|:large_blue_circle:|com/starrocks/connector/PartitionUtil.java|4|5|80.00%|[632]|
|:large_blue_circle:|com/starrocks/catalog/IcebergTable.java|10|11|90.91%|[155]|

**Review by Youngwb — APPROVED** [review]

**Pull request merged**